src/wp-includes/media.php

```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	// See which image is being returned and bail if none is found.
+	if ( ! $img = wp_get_attachment_image_src( $id, $size ) ) {
+		return false;
+	}
+
+	// Break image data into url, width, and height.
+	list( $img_url, $img_width, $img_height ) = $img;
+
+	// Bail early If we have no width or if the width set to zero.
+	if ( ! $img_width > 0 ) {
+		return false;
+	}
+
+	// Get the image meta data and bail if none is found.
+	if ( ! is_array( $img_meta = wp_get_attachment_metadata( $id ) ) ) {
+		return false;
+	}
+
+	// Build an array with image sizes.
+	$img_sizes = isset( $img_meta['sizes'] ) ? $img_meta['sizes'] : array();
+
+	// Add full size to the img_sizes array.
+	if ( isset( $img_meta['file'] ) ) {
+		$img_sizes['full'] = array(
+			'width'  => $img_meta['width'],
+			'height' => $img_meta['height'],
+			'file'   => $img_meta['file']
+		);
+
+		if ( strrpos( $img_meta['file'], '/' ) !== false ) {
+			$img_sizes['full']['file'] = substr( $img_meta['file'], strrpos( $img_meta['file'], '/' ) + 1 );
+		}
+	}
+
+	// Calculate the image aspect ratio.
+	$img_ratio = $img_height / $img_width;
+
+	// Images that have been edited in WordPress after being uploaded will
+	// contain a unique hash. We look for that hash and use it later to filter
+	// out images that are left over from previous versions.
+	// @see `wp_restore_image()`
+	$img_is_edited = preg_match( '/-e[0-9]{13}/', $img_url, $img_edit_hash );
+
+	// Set up the array of sources.
+	$sources = array();
+
+	// Loop through all available image sizes and filter out any with a
+	// different aspect ratio (crops) or are from a different edited version.
+	foreach ( $img_sizes as $img ) {
+
+		// Filter out images that are left over from previously edited versions.
+		if ( $img_is_edited && ! strpos( $img['file'], $img_edit_hash[0] ) ) {
+			continue;
+		}
+
+		// Calculate the new image ratio.
+		$img_ratio_compare = $img['height'] / $img['width'];
+
+		// If the new ratio differs by less than 0.01, use it.
+		if ( abs( $img_ratio - $img_ratio_compare ) < 0.01 ) {
+			$sources[ $img['width'] ] = path_join( dirname( $img_url ), $img['file'] ) . ' ' . $img['width'] .'w';
+		}
+	}
+
+	// Only return sources if there are more than one.
+	if ( count( $sources ) <= 1 ) {
+		return false;
+	}
+
+	/**
+	 * Filter the output of wp_get_srcset_array().
+	 *
+	 * @since 4.4.0
+	 *
+	 * @param array        $sources  An array of image urls and widths.
+	 * @param int          $id       Attachment ID for image.
+	 * @param array|string $size     Size of image, either array or string.
+	 */
+	return apply_filters( 'wp_attachment_img_srcset', $sources, $id, $size );
+}
+
+/**
+ * Get the value for the 'srcset' attribute.
+ *
+ * @since 4.4.0
+ *
+ * @param int    $id   Image attachment ID.
+ * @param string $size Optional. Name of image size. Default value: 'thumbnail'.
+ * @return string|bool A 'srcset' value string or false.
+ */
+function wp_attachment_img_srcset( $id, $size = 'thumbnail' ) {
```

Comment:
We used to use 'thumbnail', but 'medium' is a better default for these functions since 'thumbnail' will almost always return false because it's normally hard cropped. I just missed a few spots and will update this.